typescript/token/contracts/libs/FastTokenRouter.sol

```diff
@@ -0,0 +1,195 @@
+    }
+
+    /**
+     * @dev Transfers `_amountOrId` token to `_recipient` on `_destination` domain.
+     * @dev Delegates transfer logic to `_fastTransferFromSender` implementation.
+     * @dev Emits `SentTransferRemote` event on the origin chain.
+     * @param _destination The identifier of the destination chain.
+     * @param _recipient The address of the recipient on the destination chain.
+     * @param _amountOrId The amount or identifier of tokens to be sent to the remote recipient.
+     * @return messageId The identifier of the dispatched message.
+     */
+    function fastTransferRemote(
+        uint32 _destination,
+        bytes32 _recipient,
+        uint256 _amountOrId,
+        uint256 _fastFee
+    ) public payable virtual returns (bytes32 messageId) {
+        uint256 _fastTransferId = fastTransferId;
+        fastTransferId = _fastTransferId + 1;
+        bytes memory metadata = _fastTransferFromSender(
+            _amountOrId,
+            _fastFee,
+            _fastTransferId + 1
+        );
+
+        messageId = _dispatchWithGas(
+            _destination,
+            Message.format(_recipient, _amountOrId, metadata),
+            msg.value, // interchain gas payment
+            msg.sender // refund address
+        );
+        emit SentTransferRemote(_destination, _recipient, _amountOrId);
+    }
+
+    /**
+     * @dev Burns `_amount` of token from `msg.sender` balance.
+     * @dev Pays `_fastFee` of tokens to LP on source chain.
+     * @dev Returns `fastFee` as bytes in the form of metadata.
+     */
+    function _fastTransferFromSender(
+        uint256 _amount,
+        uint256 _fastFee,
+        uint256 _fastTransferId
+    ) internal virtual returns (bytes memory) {
+        _fastRecieveFrom(msg.sender, _amount);
+        return abi.encode(_fastFee, _fastTransferId);
+    }
+
+    /**
+     * @dev returns an address that indicates who should recieve the bridged tokens.
+     * @dev if _fastFees was inlcuded and someone filled the order before the mailbox made the contract call, the filler gets the funds.
+     */
+    function _getTokenRecipient(
+        address _recipient,
+        uint256 _amount,
+        uint32 _origin,
+        bytes calldata _metadata
+    ) internal view returns (address) {
+        if (_metadata.length == 0) {
+            return _recipient;
+        }
+
+        // decode metadata to extract `_fastFee` and `_fastTransferId`.
+        (uint256 _fastFee, uint256 _fastTransferId) = abi.decode(
+            _metadata,
+            (uint256, uint256)
+        );
+
+        FastTranferMetadata memory m_filledMetadata = filledFastTransfers[
+            keccak256(abi.encodePacked(_origin, _fastTransferId))
+        ];
+
+        if (
+            m_filledMetadata.fastFee <= _fastFee &&
+            _recipient == m_filledMetadata.recipient &&
+            _amount == m_filledMetadata.amount
+        ) {
+            return m_filledMetadata.filler;
+        }
```

Comment:
and then something like this that just checks if the commitment to the transfer metadata was filled already
```suggestion
        if (filledFastTransfers[fastTransferId] == fastTransferId) {
            return filler;
        } else {
            return recipient;
        }
```